### DISCUSSION TITLE
Bevy RenderPlugin User Scenarios

### DISCUSSION THREAD

**Discussion by cart** [bevyengine/bevy#2019] [Ideas]
We are in the process of reworking Bevy's renderer abstractions to improve performance (ex: via pipelining) and to better support certain scenarios like "multiple camera views", "shadows", etc. We are currently experimenting with both evolutionary changes to the current render abstractions (I'm working on that now) and revolutionary changes (adopting or building completely new renderers).

I want Bevy's rendering to support fully custom render paths, but it should also enable render plugin developers to extend "default render logic" without forcing end users to define fully custom render paths. 

The goal of this discussion is to collect requirements for "3rd party RenderPlugin" scenarios:

1. What modular logic will they export? What are the input and output render resources for that logic?
2. How will they interact with a "default render path", such as a Bevy-provided forward renderer? Each scenario is allowed to define arbitrary requirements here based on the needs of the scenario.

Once we have a representative set of scenarios, we can start thinking about the best way to accommodate them in our renderer designs.

**Comment by StarArawn** [comment]
A simple one that I can think of that illustrates how hard it is to support plugins that change rendering behavior:

Plugin Author A: 
- Writes a sky box rendering plugin. Adds a sky node as a "clear pass node"(?) to the graph.

Plugin Author B: 
- Writes a plugin to render soft point light shadows. Has a node that should run before all draw nodes. Has a custom draw pass because how else can they render the shadows with the right lighting?

Plugin Author C:
- Writes a realtime global illumination plugin has various compute passes that need to run before draw is called, has their own draw pass because again they need to render the global illumination in a way that makes sense.

User A:
- Wants to use to use the plugins from Author A, B, and C. 😢 

Implementing plugins from A, B is straight forward and same with A, C, but A, B, C is impossible as I see it without some sort of shader combiner(Think unity shader graph which kinda does this..). I don't think bevy should accommodate this though at least not in the near future..

**Comment by mtsr** [comment]
I'll follow @StarArawn's format, unless somebody proposes something better.

Plugin Author A:
* Writes a bloom post process effect. For HDR this requires render textures from previous passes to be in Rgba16Float or similar.

Plugin Author B:
* Writes a SuperSampling AA post process effect. Effectively downscaling the render textures from previous passes, so requiring those to be rendered at a higher than window resolution.

Plugin Author C:
* Writes a TAA plugin. This requires access to render textures from previous frames in addition to current frame.

User A:
* Writes an single pixel outline post process effect. This requires access to color, depth and normals, but has to happen either without MSAA or (more complicated) after any MSAA has happened. The latter is harder because it requires depth to be available after resolve.

User B:
* Writes an integer upscale post process effect. While this can be done with a simple fullscreen shader, for best performance, it should be possible to render at different resolutions before and after the effect, because otherwise a lot of expensive pixels from before upscaling will be discarded.

Other effects that might be interesting:
* Screen Space Ambient Occlusion
* Image Based Lighting (either from existing images, or from rendering other views)
* Order Independent Transparency
* Skinning (for animations)
* Cloth simulation

**Comment by mtsr** [comment]
This page was linked on discord, which is probably a good reference https://docs.unity3d.com/Packages/com.unity.render-pipelines.high-definition@10.4/manual/Custom-Pass-Injection-Points.html

It describes this diagram: ![Unity custom pass injection points](https://docs.unity3d.com/Packages/com.unity.render-pipelines.high-definition@10.4/manual/images/HDRP-frame-graph-diagram.png)
